tests/test_wfc/test_wfc_solver.py

```diff
@@ -0,0 +1,157 @@
+        return 1, 2
+
+    def patHeu(weights: NDArray[np.bool_], wave: NDArray[np.bool_]) -> int:
+        assert np.array_equal(weights, my_wave[:, 1, 2])
+        return 3
+
+    assert wfc_solver.observe(
+        my_wave, locationHeuristic=locHeu, patternHeuristic=patHeu
+    ) == (
+        3,
+        1,
+        2,
+    )
+
+
+def test_propagate() -> None:
+    wave = np.ones((3, 3, 4), dtype=bool)
+    adjLists = {}
+    # checkerboard #0/#1 or solid fill #2
+    adjLists[(+1, 0)] = adjLists[(-1, 0)] = adjLists[(0, +1)] = adjLists[(0, -1)] = [
+        [1],
+        [0],
+        [2],
+    ]
+    wave[:, 0, 0] = False
+    wave[0, 0, 0] = True
+    adj = wfc_solver.makeAdj(adjLists)
+    wfc_solver.propagate(wave, adj, periodic=False)
+    expected_result = np.array(
+        [
+            [
+                [True, False, True, False],
+                [False, True, False, True],
+                [True, False, True, False],
+            ],
+            [
+                [False, True, False, True],
+                [True, False, True, False],
+                [False, True, False, True],
+            ],
+            [
+                [False, False, False, False],
+                [False, False, False, False],
+                [False, False, False, False],
+            ],
+        ]
+    )
+    assert np.array_equal(wave, expected_result)
+
+
+def test_run() -> None:
+    wave = wfc_solver.makeWave(3, 3, 4)
+    adjLists = {}
+    adjLists[(+1, 0)] = adjLists[(-1, 0)] = adjLists[(0, +1)] = adjLists[(0, -1)] = [
+        [1],
+        [0],
+        [2],
+    ]
+    adj = wfc_solver.makeAdj(adjLists)
+
+    first_result = wfc_solver.run(
+        wave.copy(),
+        adj,
+        locationHeuristic=wfc_solver.lexicalLocationHeuristic,
+        patternHeuristic=wfc_solver.lexicalPatternHeuristic,
+        periodic=False,
+    )
+
+    expected_first_result = np.array([[0, 1, 0, 1], [1, 0, 1, 0], [0, 1, 0, 1]])
+
+    assert np.array_equal(first_result, expected_first_result)
+
+    event_log: list = []
+
+    def onChoice(pattern: int, i: int, j: int) -> None:
+        event_log.append((pattern, i, j))
+
+    def onBacktrack() -> None:
+        event_log.append("backtrack")
+
+    second_result = wfc_solver.run(
+        wave.copy(),
+        adj,
+        locationHeuristic=wfc_solver.lexicalLocationHeuristic,
+        patternHeuristic=wfc_solver.lexicalPatternHeuristic,
+        periodic=True,
+        backtracking=True,
+        onChoice=onChoice,
+        onBacktrack=onBacktrack,
+    )
+
+    expected_second_result = np.array([[2, 2, 2, 2], [2, 2, 2, 2], [2, 2, 2, 2]])
+
+    assert np.array_equal(second_result, expected_second_result)
+    print(event_log)
+    assert event_log == [(0, 0, 0), "backtrack", (2, 0, 0)]
+
+    class Infeasible(Exception):
+        pass
+
+    def explode(wave: NDArray[np.bool_]) -> bool:
+        if wave.sum() < 20:
+            raise Infeasible
+        return False
+
+    try:

```

Comment:
Can we not just call `wfc_solver.run` as then the exception will be printed rather than the `assert happy` at the end